File: Breadth_first_search_algorithm/Bài_toán_10.py
from collections import deque

def bfs_general(graph):
    n = len(graph)
    visited = set()
    components = []

    for start in range(n):
        if start not in visited:
            queue = deque([start])
            component = []
            while queue:
                node = queue.popleft()
                if node not in visited:
                    visited.add(node)
                    component.append(node)
                    for neighbor in graph[node]:
                        if neighbor not in visited:
                            queue.append(neighbor)
            components.append(component)
    return components

graph = {
    0: [1, 1],
    1: [0, 2],
    2: [1, 2],
    3: [],
    4: [5],
    5: [4]
}

print("All BFS components:")
for i, comp in enumerate(bfs_general(graph)):
    print(f"Component {i+1}:", comp)

File: Breadth_first_search_algorithm/Bài_toán_10.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

void bfs_component(int start, const vector<vector<int>>& adj, vector<bool>& visited, vector<int>& comp) {
    queue<int> q;
    q.push(start);
    visited[start] = true;

    while (!q.empty()) {
        int u = q.front(); q.pop();
        comp.push_back(u);
        for (int v : adj[u]) {
            if (!visited[v]) {
                visited[v] = true;
                q.push(v);
            }
        }
    }
}

void bfs_general(const vector<vector<int>>& adj) {
    int n = adj.size();
    vector<bool> visited(n, false);
    int component_id = 1;

    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            vector<int> comp;
            bfs_component(i, adj, visited, comp);
            cout << "Component " << component_id++ << ": ";
            for (int v : comp) cout << v << " ";
            cout << "\n";
        }
    }
}

int main() {
    vector<vector<int>> graph = {
        {1, 1},
        {0, 2},
        {1, 2},
        {},
        {5},
        {4}
    };

    cout << "All BFS components:\n";
    bfs_general(graph);

    return 0;
}
